base/pkg/read.jl

```diff
@@ -2,15 +2,70 @@
 
 module Read
 
-import ...LibGit2, ..Cache, ..Reqs, ...Pkg.PkgError, ..Dir
+import ...LibGit2, ..Cache, ..Reqs, ...Pkg.PkgError, ..Dir, ..isdevmetadata
 using ..Types
 
+devmetapkgpath(s::AbstractString) = joinpath(uppercase(string(s[1])), s * ".versions")
+
 readstrip(path...) = strip(readstring(joinpath(path...)))
 
-url(pkg::AbstractString) = readstrip(Dir.path("METADATA"), pkg, "url")
-sha1(pkg::AbstractString, ver::VersionNumber) = readstrip(Dir.path("METADATA"), pkg, "versions", string(ver), "sha1")
+function url(pkg::AbstractString, dev::Bool = false)
+    if dev
+        chomp(readline(joinpath(Dir.path("METADATA"), devmetapkgpath(pkg))))
+    else
+        readstrip(Dir.path("METADATA"), pkg, "url")
+    end
+end
+
+# Should only be called when cache is guaranteed to be updated
```

Comment:
An open question is whether a package registry should have a file per package or a single large file with information about all packages or something in between. We've already hit the limit where GitHub doesn't like the number of files in METADATA (1000), but real file systems are ok with this amount. If you use a single file then a major disadvantage is that it's hard/impossible to do lazy parsing – you have to parse everything at once. We could shard the package list (by first letter?), or provide a mechanism for finding where in a single registry file each package lives so that we can parse just that section.